google_api_lib/tasks.py

```diff
@@ -620,51 +620,60 @@ def update_active_users(self, hunt_id):
         else:
             body["pageToken"] = response["nextPageToken"]
 
-    with transaction.atomic():
-        user_puzzle_pairs = [
-            f"{user_pk}-{puzzle_pk}" for (user_pk, puzzle_pk) in latest.keys()
-        ]
-        old_activities = (
-            PuzzleActivity.objects.annotate(
-                user_puzzle_pair=Concat(
-                    F("user_id"), Value("-"), F("puzzle_id"), output_field=CharField()
+    try:
+        with transaction.atomic():
+            user_puzzle_pairs = [
+                f"{user_pk}-{puzzle_pk}" for (user_pk, puzzle_pk) in latest.keys()
+            ]
+            old_activities = (
+                PuzzleActivity.objects.annotate(
+                    user_puzzle_pair=Concat(
+                        F("user_id"),
+                        Value("-"),
+                        F("puzzle_id"),
+                        output_field=CharField(),
+                    )
                 )
+                .filter(user_puzzle_pair__in=user_puzzle_pairs)
+                .select_for_update()
+            )
+            old_keys = [(e.user_id, e.puzzle_id) for e in old_activities]
+
+            PuzzleActivity.objects.bulk_create(
```

Comment:
DB `bulk_create` and `bulk_update` operations are pretty susceptible to timeout / OOM issues.

What is the scale of `latest.items()`. If it can exceed O(1000), might be worth chunking this.

Reply:
It depends on how often the CRON job is set to run. At the intervals we set, I think its like O(10).